File: diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int max;

    int helper(TreeNode* node){
        if(!node)
            return 0;
        
        int left = helper(node->left);
        int right = helper(node->right);

        if(left + right > max)
            max = left + right;

        return std::max(left, right) + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        max = 0;
        helper(root);
        return max;
    }
};

File: diameter-of-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private int max;

    private int Helper(TreeNode node){
        if(node == null)
            return 0;
        
        int left = Helper(node.left);
        int right = Helper(node.right);

        if(left + right > max)
            max = left + right;

        return Math.Max(left, right) + 1;
    }

    public int DiameterOfBinaryTree(TreeNode root) {
        max = 0;
        Helper(root);
        return max;
    }
}